### PR TITLE
Governance fix data download

### LINKED ISSUE #1807: [governance] Update proposal list results after user votes
After a user successfully votes the proposal results that are used for the list should be updated the same as the transaction details one is as well.

### PR DESCRIPTION
**Pull request by vctt94**
~~This PR is built on top of https://github.com/decred/decrediton/pull/2192.~~

When testing this PR if one wants to cast a vote, it will be necessary to have eligible tickets.

It closes https://github.com/decred/decrediton/issues/2076, closes https://github.com/decred/decrediton/issues/2101, closes https://github.com/decred/decrediton/issues/2047, closes https://github.com/decred/decrediton/issues/2015 (duplicated), and closes https://github.com/decred/decrediton/issues/2166, and closes #1807

I am caching the eligible tickets and as @lukebp suggested we are only showing cast vote options for active voting proposals.

I believe we can also cache the information of other votes (we have this information when querying for the proposal details) and show votes of proposals which we have at the finished vote tab, but I believe this is open to discussion. 

We could also close https://github.com/decred/decrediton/issues/2042, but for that we need commit https://github.com/decred/politeia/commit/8b25fa2f44799c4ee786d04da4bfa450ed255e66 deployed. I left a TODO in the code: https://github.com/decred/decrediton/compare/master...vctt94:governance-fix-data-download?expand=1#diff-c8399d09f1bddd336993b884a3e8d512R31 and after it gets deployed will be no problem fixing that.

### CODE REVIEW

File: app/components/views/GovernancePage/Proposals/Details/Page.js [hunk 1]
```diff
@@ -14,38 +14,45 @@ export default ({ viewedProposalDetails, goBackHistory,
   showPurchaseTicketsPage, hasTickets, onVoteOptionSelected, onUpdateVoteChoice,
   newVoteChoice, updateVoteChoiceAttempt, tsDate, text }) =>
 {
-  const { name, token, hasEligibleTickets, voteStatus, voteOptions,
-    voteCounts, creator, timestamp, endTimestamp, currentVoteChoice,
+  const { name, token, voteStatus, proposalStatus, voteOptions, voteCounts,
+    creator, timestamp, endTimestamp, currentVoteChoice, hasEligibleTickets,
     version, quorumMinimumVotes } = viewedProposalDetails;
-  const eligibleTicketCount = viewedProposalDetails.eligibleTickets.length;
+  const eligibleTicketCount = viewedProposalDetails.eligibleTicketCount;
```
Comment (matheusd) [inline app/components/views/GovernancePage/Proposals/Details/Page.js]: Given you changed `eligibleTickets` to be _all_ eligible tickets (vs the ones from the wallet) the message is showing a large number of tickets for voting (vs before when it showed the number of tickets _from the wallet_ that are eligible for voting).

File: app/main_dev/paths.js [hunk 2]
```diff
@@ -100,28 +87,159 @@ export function getExecutablePath(name, custombinpath) {
...
+// saveEligibleTickets receives a proposal token and its eligible tickets object.
+// it checks if proposal path exists, creates it if it does not exist.
+// and write a eligibletickets.json file, with { eligibleTickets: [tickets]) }
+export function saveEligibleTickets (token, eligibleTickets) {
+  let proposalPath = getProposalPathFromPoliteia(token);
+  if (!fs.pathExistsSync(proposalPath)) {
+    proposalPath = setPoliteiaProposalPath(token);
+  }
+  const fullPath = path.join(proposalPath, "eligibletickets.json");
+  fs.writeFile(fullPath, JSON.stringify(eligibleTickets));
```
Comment (matheusd) [inline app/main_dev/paths.js]: Use `{ mode: 0o600 }` as a small security protection.

File: app/main_dev/paths.js [hunk 3]
```diff
@@ -100,28 +87,159 @@ export function getExecutablePath(name, custombinpath) {
...
+  return path.join(getAppDataDirectory(), "politeia");
+}
+
+// setPoliteiaPath sets the politeia path which proposals are cached.
+export function setPoliteiaPath () {
+  const politeiaPath = getPoliteiaPath();
+  if (fs.pathExistsSync(politeiaPath)) {
+    return;
+  }
+  fs.mkdirSync(politeiaPath);
```
Comment (matheusd) [inline app/main_dev/paths.js]: Use `{ mode: 0o700 }` as a small security protection.

File: app/main_dev/paths.js [hunk 4]
```diff
@@ -100,28 +87,159 @@ export function getExecutablePath(name, custombinpath) {
...
+}
+
+// setPoliteiaProposalPath mkdir if directory of proposal does not exists.
+export function setPoliteiaProposalPath (token) {
+  let proposalPath = getProposalPathFromPoliteia(token);
+  if (fs.pathExistsSync(proposalPath)) {
+    return;
+  }
+  proposalPath = path.join(getPoliteiaPath(), token);
+  fs.mkdirSync(proposalPath);
```
Comment (matheusd) [inline app/main_dev/paths.js]: Use `{ mode: 0o700 }`

File: app/main_dev/paths.js [hunk 5]
```diff
@@ -100,28 +87,159 @@ export function getExecutablePath(name, custombinpath) {
...
+}
+
+// getWalletPiPath gets the wallet politeia path if it exists, otherwise
+// it creates its path and returns it.
+function getWalletPiPath (testnet, walletName) {
+  const walletPiPath = path.join(getWalletPath(testnet, walletName), "politeia");
+  if (fs.pathExistsSync(walletPiPath)) {
+    return walletPiPath;
+  }
+  fs.mkdirSync(walletPiPath);
```
Comment (matheusd) [inline app/main_dev/paths.js]: Use `{ mode 0o700 }`

File: app/main_dev/paths.js [hunk 6]
```diff
@@ -100,28 +87,159 @@ export function getExecutablePath(name, custombinpath) {
...
+
+// savePiVote checks if proposal directory exists, creates it, otherwise and
+// creates the vote.json file with the vote information.
+// we do not delete this directory after, as we use it to check for finished
+// votings with the wallet.
+export function savePiVote (vote, token, testnet, walletName) {
+  const walletPath = getWalletPiPath(testnet, walletName);
+  const proposalPath = path.join(walletPath, token);
+  if (!fs.pathExistsSync(proposalPath)) {
+    fs.mkdirSync(proposalPath);
```
Comment (matheusd) [inline app/main_dev/paths.js]: Use `{ mode: 0o700 }`

File: app/main_dev/paths.js [hunk 7]
```diff
@@ -100,28 +87,159 @@ export function getExecutablePath(name, custombinpath) {
...
+  }
+  const fullPath = path.join(proposalPath, "vote.json");
+  fs.writeFile(fullPath, JSON.stringify(vote));
```
Comment (matheusd) [inline app/main_dev/paths.js]: Use `{ mode: 0o600 }`

**Comment by lukebp** [comment]
The Politeia changes that this needs should be deployed to mainnet towards the end of this week or early next week.

**Comment by vctt94** [comment]
Good catch on the mode type when writing files or directory. I fixed them. 

Now that the new politeia code is released in mainnet I believe it is fine for another review.

I made some tests in `mainnet` and find no issues. I did not vote, though, as we do not have voting proposals in mainnet. Voting in testnet is working.

**Comment by vctt94** [comment]
That is what I mentioned I believe it is open to debate.

After voting, the information will keep cached until the user lose its information. So we can simply show it.

For proposals which already finished voting or the user has voted but has not cached, we will need more steps to save this information.

The proposal details end-point: 
https://github.com/decred/politeia/blob/master/politeiawww/api/www/v1/api.md#proposal-details
 does not have the cast votes for a specific proposals.

So we need to call cast votes and check for eligible tickets from the wallet and then cache this information.

The problem with that, is that proposals which a user has voted not necessarily will show its votes, as the user may have not opened it. But I believe this inconsistency is fine, right?

I can add that to this PR, but I can also include that working on https://github.com/decred/decrediton/issues/2244. Which one you prefer.

**Pull request merged**